src/lem/constrainer.rs

```diff
@@ -90,28 +148,226 @@ impl LEM {
         }
     }
 
+    fn z_ptr_from_valuation<F: LurkField>(
+        branch_path_info: &Option<Boolean>,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(branch_path_info)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn stack_hash_slots<F: LurkField>(
+        branch_path_info: &Option<Boolean>,
```

Comment:
`Option<Boolean>` makes me wonder if there is a better formulation hiding in here. Just thinking out loud, not making a definite statement.